File: android/core/domain/src/main/java/kr/co/cdd/payboard/core/domain/usecase/BillingDateCalculator.kt
package kr.co.cdd.payboard.core.domain.usecase

import kr.co.cdd.payboard.core.domain.model.BillingCycle
import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime

object BillingDateCalculator {
    fun advance(from: Instant, cycle: BillingCycle, zoneId: ZoneId = ZoneId.systemDefault()): Instant {
        val zoned = ZonedDateTime.ofInstant(from, zoneId)
        return when (cycle) {
            BillingCycle.Monthly -> zoned.plusMonths(1).toInstant()
            BillingCycle.Yearly -> zoned.plusYears(1).toInstant()
            is BillingCycle.CustomDays -> zoned.plusDays(cycle.days.toLong().coerceAtLeast(1)).toInstant()
        }
    }
}

File: ios/Sources/Domain/BillingDateCalculator.swift
import Foundation

public enum BillingDateCalculator {
    public static func advance(from date: Date, cycle: BillingCycle, calendar: Calendar = .current) -> Date {
        switch cycle {
        case .monthly:
            return calendar.date(byAdding: .month, value: 1, to: date) ?? date
        case .yearly:
            return calendar.date(byAdding: .year, value: 1, to: date) ?? date
        case let .customDays(days):
            return calendar.date(byAdding: .day, value: max(1, days), to: date) ?? date
        }
    }
}
